pymecht/MatModel.py

```diff
@@ -1044,4 +1044,136 @@ def _energy_stress(self,A,B,mean_theta,sigma,aniso_fraction,**extra_args):
+        
+        # Construct a string of the symbols to define as such without needing
+        # numerical values
+        param_names_string = ""
+        for name in self.param_names:
+            param_names_string += name + " "
+        
+        other_symbols = sp.symbols(param_names_string)
+        
+        SEDF = sp.expand(self.energy_form)
+        
+        # Take symbolic derivatives
+        dSEDFdI =   sp.diff(SEDF,I1), \
+                    sp.diff(SEDF,I2), \
+                    sp.diff(SEDF,I3), \
+                    sp.diff(SEDF,I4)
+        
+        # Store derivatives as strings
+        dSEDFdI = [str(sp.powdenest(sp.factor_terms(sp.powsimp(i)))) for i in dSEDFdI]
+        
+        # Format derivatives for consistent syntax
+        dSEDFdI = self.replaceSympySyntax(dSEDFdI)
+        
+        # Update attribute of class.
+        self.denergy_formdI = dSEDFdI
+
+        # Calculate the partial derivatives wrt the linear and nonlinear params
+        if dparams == True:
+            dSEDFdparams = []
+            for param in self.param_names:
+                # if param[:6] == "ltheta"
+                #     dSEDFdparams +=
+                exec(param + " = sp.symbols(param)")
+                dSEDFdparams += [sp.diff(SEDF,eval(param))]
+            dSEDFdparams = [str(sp.powsimp(i)) for i in dSEDFdparams]
+            dSEDFdparams = self.replaceSympySyntax(dSEDFdparams)
+            self.denergy_formdparams = dSEDFdparams
+        return
+
+    def replaceSympySyntax(self, function):
+        function = [i.replace("I1","self.I1") for i in function]
+        function = [i.replace("I2","self.I2") for i in function]
+        function = [i.replace("I3","self.I3") for i in function]
+        function = [i.replace("I4","self.I4") for i in function]
+        function = [i.replace("exp","np.exp") for i in function]
+        function = [i.replace("sqrt","np.sqrt") for i in function]
+        return function
+
+    def _energy(self,**extra_args):
+        # WARNING: UNTESTED
+        print("Warning: Arbitrary MatModel has not been tested. Proceed with caution!")
+        for name in self.param_names:
+            exec(name + "=" + "%s" %(extra_args[name]))
+        
+        SEDF = self.energy_form.replace("sp","np")
+        return eval(SEDF)
+
+    def partial_deriv(self,**extra_args):
+        # Evaluate the numerical values of the partial derivatives from the
+        # strings in self. Need some funky syntax to allow eval() to have
+        # access to the variable names in extra_args. It unpacks these values
+        # into a new dictionary, along with self and np.
+        dSEDFdI = [eval(i, {"self": self, "np": np,**extra_args}) for i in self.denergy_formdI]
+        # Not been able to cast this using list comprehension. It sets unused
+        # entries to be None.
+        for i in range(len(dSEDFdI)):
+            if "I%s"%(i+1) not in self.energy_form:
+                dSEDFdI[i] = None
+        
```

Comment:
Here we can do sum of the first three terms as follows:

```
elif i<3:
   dSEDFdI[i] = np.sum(dSEDFdI[i])
```

That should avoid running into the #1 issue that you raised a long time ago.